### DISCUSSION TITLE
Built-in light detection for when blend mode is lightonly

### DISCUSSION THREAD

**Discussion by almighty00ZEDD** [godotengine/godot-proposals#11029] [Engine Core]
Hello,

Ever made a sprite's canvas material to blend mode lightonly? it does it's job pretty well by only revealing it when light hits it!

The missing functionality i wish the godot team would add is some signal for when the sprite is being revealed or not to use it as light detection, because what everyone does at the moment is using area2D and raycast which can slow the performance if the shapes are not circular...

So i checked the C++ code and my intuition was right, the core function for revealing the sprite based on light is just there, and just needs some extra code to bind a bool variable for the GDscript api to be used as a light detection signal!

This can save a lot of time and ressources especially if your game like mine is heavily based on light.

Thanks.

**Comment by Calinou** [comment]
The bool property will be `true` as soon as the sprite's AABB touches the light's AABB ever so slightly. This means that it can be `true` even when the sprite isn't actually visible to the player, depending on the sprite's and light's shape.

This is OK as an approximation, but it might not be precise enough for some use cases as it'll return `true` too early.

**Reply by almighty00ZEDD** [reply]
ow, what about taking the alpha channel as a parameter as not setting to true as long as the texture part is actually drawn on (ignore when the original alpha is 0.0)?

i didn't take a deep dive into the actual c++ code but i just know there's a way, since the light() section in the shader isn't made for 2D and the shaders themselves are a one way interactive as i tried comparing the original TEXTURE color with the current one but no real way to know if this worked as having a script check for a bool in a shader is straight impossible (or just didn't figure it out yet).

For now i used area2D (non regular shape for the flashlight) and a raycast2D to check for obstacles but it feels like too much processing for a potential built-in feature.
